_posts/2021-04-15-testing-react-tracking-with-jest-and-enzyme.markdown

```diff
@@ -0,0 +1,242 @@
+
+Our use-case was a button that would navigate the user to a new page. The button had been implemented in a previous
+PR, and now we wanted to make it execute a tracking call before navigating.
+
+We use Segment for tracking, and their tracking setup relies on a JS snippet being available on your pages. That
+snippet sets a `window.analytics` property, which in turn
+[has a](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#track) `.track()`
+method. On a fundamental level, all of our tracking calls boil down to a call to `window.analytics.track()`, where
+we pass a list of properties to `.track()`
+
+# Adding react-tracking
+
+First, there was a bit of setup required to get react-tracking working. The react-tracking package
+[assumes you're using Google Tag Manager by default](https://github.com/NYTimes/react-tracking#custom-optionsdispatch-for-tracking-data),
+but allows you to override that behavior with a custom `dispatch` function. In our case, we wrap our React apps in
+a `<BaseApp>` component, so we added a new `<TrackerContextCohesion>` component with a custom `dispatch` that would
+be available to all of our React apps:
+
+```ts
+interface TrackEventProps {
+  data: { [key: string]: string }
+  options: {}
+  callback: () => void
+}
+
+export const trackEvent = ({ data, options, callback }: Partial<TrackEventProps>): void => {
+  // Action can be something like "click" or "Viewed tooltip"; we pull it out and send it to Segment
+  // separately since it defines the event type
+  const actionName = data.action || data.action_type
+  const trackingData = omit(data, ["action_type", "action"])
+
+  if (actionName) {
+    // This is where we actually call Segment and pass the tracking event
+    window.analytics.track(actionName, trackingData, options, callback)
+  } else {
+    console.error(`Unknown analytics schema being used: ${JSON.stringify(data)}`)
+  }
+}
+
+// We're following the instructions from react-tracking's README on overriding the dispatch function
+export const TrackerContextCohesion = track(
+  {},
+  {
+    dispatch: (args) => {
+      trackEvent(args)
+    },
+  }
+)((props) => {
+  return props.children
+})
+```
+
+This allows us to make tracking calls in our components, including the passing of custom callback functions:
+
+```ts
+const { trackEvent } = useTracking()
+
+// This type is defined in Cohesion, making it easy for us to ensure we have the correct tracking
+// properties
+const trackingArgs: ClickedEditArtwork = {
+  action: ActionType.clickedEditArtwork,
+  context_module: ContextModule.toDoList,
+  context_page_owner_type: OwnerType.home,
+  destination_page_owner_id: internalID,
+  destination_page_owner_slug: slug,
+  destination_page_owner_type: OwnerType.artwork,
+  destination_path: "/artworks",
+  label: "Add info",
+}
+
+const handleAddInfoClick = (): void => {
+  trackEvent({
+    data: trackingArgs,
+    options: {},
```

Comment:
Maybe fill in a couple options just so this param doesn't hang, and perhaps add an in-code comment linking out to the docs,